public/app/core/components/connectWithCleanUp.tsx

```diff
@@ -0,0 +1,36 @@
+import { MapStateToPropsParam, MapDispatchToPropsParam, connect, useDispatch } from 'react-redux';
+import { StateSelector, cleanUpAction } from '../actions/cleanUp';
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import hoistNonReactStatics from 'hoist-non-react-statics';
+
+export const connectWithCleanUp = <
```

Comment:
I don't think I follow your comment @aocenas

Reply:
I meant that we should probably have `useConnectWithCleanUp` hook that is then just used in `connectWithCleanUp` HOC. That way we can use the same functionality both ways. Even though arguably you can use HOC with functional components it is bit more awkward to build on top of HOCs while hooks can be more easily reuse in other hooks.

I did something like that [here](https://github.com/grafana/grafana/blob/b0d6ac5f42c0a082d77fb5fa256ef885e1e828e9/public/app/features/explore/useLiveTailControls.ts#L54) though that uses render prop component instead of HOC.

Reply:
I think we could do that when the need for cleaning up Functional Components that use state arises. Don't know of any such case today.